connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```
@@ -1320,89 +1317,192 @@ void alterSinkConnectorOffsets(String connName, Connector connector, Map<String,
                     }
 
-                    Set<TopicPartition> partitionsToReset = parsedOffsets.entrySet()
-                            .stream()
-                            .filter(entry -> entry.getValue() == null)
-                            .map(Map.Entry::getKey)
-                            .collect(Collectors.toSet());
-
-                    if (!partitionsToReset.isEmpty()) {
-                        log.debug("Deleting the consumer group offsets for the following topic partitions using an admin client for sink connector {}: {}.",
-                                connName, partitionsToReset);
-                        DeleteConsumerGroupOffsetsOptions deleteConsumerGroupOffsetsOptions = new DeleteConsumerGroupOffsetsOptions().timeoutMs(
-                                (int) ConnectResource.DEFAULT_REST_REQUEST_TIMEOUT_MS);
-                        DeleteConsumerGroupOffsetsResult deleteConsumerGroupOffsetsResult = admin.deleteConsumerGroupOffsets(groupId, partitionsToReset,
-                                deleteConsumerGroupOffsetsOptions);
-
-                        adminFutures.add(deleteConsumerGroupOffsetsResult.all());
+                    boolean alterOffsetsResult;
+                    try {
+                        alterOffsetsResult = ((SinkConnector) connector).alterOffsets(connectorConfig, offsetsToWrite);
+                    } catch (UnsupportedOperationException e) {
+                        log.error("Failed to modify offsets for connector {} because it doesn't support external modification of offsets",
+                                connName, e);
+                        throw new ConnectException("Failed to modify offsets for connector " + connName + " because it doesn't support external " +
+                                "modification of offsets", e);
                     }
+                    updateTimerAndCheckExpiry(timer, "Timed out while calling the 'alterOffsets' method for sink connector " + connName);
```

Should we add a note to `SourceConnector::alterOffsets` and `SinkConnector::alterOffsets` that this method may be invoked multiple times with the same arguments if an external operation (writing to the offsets topic, deleting a consumer group, etc.) fails? That way, developers can know to silently ignore requests to reset offsets that don't appear to be present based on the connector's own external tracking logic.

---

Hm good point, we really don't want to get into a situation where let's say a call to `SinkConnector::alterOffsets` succeeds but the call to delete the consumer group fails and then any subsequent external requests to delete the connector's offsets fails because the `SinkConnector::alterOffsets` method bails out..

On the other hand, I'm also concerned about making the Javadoc for those methods overly complicated. Furthermore, since we'll be calling `alterOffsets` on separate `Connector` instances, it's not like they can track state between multiple calls. Any suggestions on the wording so that we can maintain the right balance? Maybe we could add a hint that offsets that appear non-existent should be handled gracefully if possible (this also covers the case where an alter offsets request is used to skip over some data for sink connectors)?

---

> Maybe we could add a hint that offsets that appear non-existent should be handled gracefully if possible (this also covers the case where an alter offsets request is used to skip over some data for sink connectors)?

I like this, but I wish we could be a little more explicit. I think ideally we want connectors to respond with a no-op for resets of partitions that don't appear to exist, and for non-resets (assuming they do external offset tracking at all) either store that partition/offset pair externally, or throw an exception to indicate that the operation cannot be performed.